projects/bootloader/scripts/can_datagram.py

```diff
@@ -0,0 +1,294 @@
+                          len(self._node_ids),
+                          *(self._node_ids),
+                          len(self._data) & 0xff,
+                          (len(self._data) >> 8) & 0xff,
+                          *(self._data)])
+
+    # Accessors and mutators for the datagram
+
+    @property
+    def protocol_version(self):
+        """This function describes the protocol version property."""
+        return self._protocol_version
+
+    @protocol_version.setter
+    def protocol_version(self, value):
+        """This function sets the protocol version."""
+        assert value & 0xff == value
+        self._protocol_version = value & 0xff
+
+    @property
+    def datagram_type_id(self):
+        """This function describes the datagram type id property."""
+        return self._datagram_type_id
+
+    @datagram_type_id.setter
+    def datagram_type_id(self, value):
+        """This function sets the datagram type id."""
+        assert value & 0xff == value
+        self._datagram_type_id = value & 0xff
+
+    @property
+    def node_ids(self):
+        """This function describes the node ids property."""
+        return self._node_ids
+
+    @node_ids.setter
+    def node_ids(self, value):
+        """This function sets the node ids."""
+        assert isinstance(value, list)
+        for val in value:
+            assert val & 0xff == val
+            assert val < 64
+        self._node_ids = value
+
+    @property
+    def data(self):
+        return self._data
+
+    @data.setter
+    def data(self, value):
+        """This function sets the data."""
+        assert isinstance(value, bytearray)
+        self._data = value
+
+    def _check_kwargs(self, **kwargs):
+        """This function checks that all variables are as expected"""
+
+        args = [
+            "datagram_type_id",
+            "node_ids",
+            "data"]
+
+        # Check all arguments are present
+        for arg in args:
+            assert arg in kwargs
+
+        # Check that types are as expected
+        assert not isinstance(kwargs["datagram_type_id"], list)
+        assert isinstance(kwargs["node_ids"], list)
+        assert isinstance(kwargs["data"], bytearray)
+
+        # Verify all inputs
+        assert kwargs["datagram_type_id"] & 0xff == kwargs["datagram_type_id"]
+
+    def _convert_to_bytearray(self, in_value, bytes):
+        """This is a helper function that creates a little-endian bytearray"""
+        out_bytearray = bytearray()
+        for i in range(0, bytes):
+            out_bytearray.append((in_value >> (8 * i)) & 0xff)
+        return out_bytearray
+
+    def _convert_from_bytearray(self, in_bytearray, bytes):
+        """This is a helper function that converts a bytearray into a value"""
+        value = 0
+        for i in range(0, bytes):
+            value = value | ((in_bytearray[i] & 0xff) << (i * 8))
+        return value
+
+    def _calculate_crc32(self, datagram_type_id, node_ids, data):
```

Comment:
Python point: you don't use `self` in any of these, and you call them like `cls._convert_to_bytearray(cls, ...)`. Instead you could decorate them with `@staticmethod`, so that you can omit the `self` argument and call them like `cls._convert_to_bytearray(...)`. That's more Pythonic.

Reply:
I will make this change, but this makes me very curious-- do you think you could explain to me on the side about pythonic class methods? e.g. when it is appropriate to use classmethod, staticmethod, and no decorator? Thanks!

Reply:
There's sort of two distinctions here: one between no decorator and using classmethod or staticmethod, and then one between classmethod and staticmethod.

Methods that belong to a particular instance of a class shouldn't be decorated with classmethod or staticmethod, and methods that belong to the class as a whole (not to any particular instance, just grouped with the class so you can call it with `MyClass.my_method`) should be decorated with classmethod / staticmethod. It's the same as the distinction between static and instance methods in C++/Java/C# if you're familiar with those.

Within 'static' methods, a method decorated with classmethod will receive a `cls` parameter as the first argument, like an instance method gets `self` as the first argument. `cls` is an object representing the class (everything is an object in Python, even classes), and in particular you can call `cls()` to construct an object of that class. It's like you did `cls = MyClass`. The main use for this is so that if you call a class method on a subclass, it'll automatically use the subclass in `cls` and you don't have to rewrite the class method. Classmethods are pretty much a Python-only thing, I haven't seen them elsewhere.

A method decorated with staticmethod, on the other hand, is very much like a C++/Java/C# `static` method. It doesn't receive any special parameters, and it's static so it can't access instance fields. It's basically a regular function except grouped with a class so you can do `MyClass.my_static_method()` to call it.